c10/core/impl/SizesAndStrides.h

```diff
@@ -17,26 +30,118 @@ class C10_API SizesAndStrides {
   using strides_iterator = int64_t*;
   using strides_const_iterator = const int64_t*;
 
-  SizesAndStrides() : sizes_{0}, strides_{1} {}
+  SizesAndStrides() {
+    setInlineSize(1);
+    size_at_unchecked(0) = 0;
+    stride_at_unchecked(0) = 1;
+  }
+
+  ~SizesAndStrides() {
+    if (C10_UNLIKELY(!isInline())) {
+      delete outOfLineStorage();
+    }
+  }
+
+  SizesAndStrides(const SizesAndStrides& rhs) {
+    if (C10_LIKELY(rhs.isInline())) {
+      copyFromInline(rhs);
+    } else {
+      taggedStorageOrSize_ = reinterpret_cast<int64_t>(new std::vector<int64_t>(*rhs.outOfLineStorage()));
+    }
+  }
+
+  SizesAndStrides& operator=(const SizesAndStrides& rhs) {
+    if (this == &rhs) {
+      return *this;
+    }
+    if (C10_LIKELY(rhs.isInline())) {
+      if (C10_UNLIKELY(!isInline())) {
+        delete outOfLineStorage();
+      }
+      copyFromInline(rhs);
+    } else {
+      if (isInline()) {
+        taggedStorageOrSize_ = reinterpret_cast<int64_t>(new std::vector<int64_t>(*rhs.outOfLineStorage()));
+      } else {
+        // Both out of line. Copy their storage into ours.
+        *outOfLineStorage() = *rhs.outOfLineStorage();
+      }
+    }
+    return *this;
+  }
+
+  // Move from rhs. rhs.size() == 0 afterwards.
+  SizesAndStrides(SizesAndStrides&& rhs) : taggedStorageOrSize_(rhs.taggedStorageOrSize_) {
+    if (C10_LIKELY(isInline())) {
+      memcpy(inlineStorage_, rhs.inlineStorage_, sizeof(inlineStorage_));
+    }
+
+    rhs.setInlineSize(0);
+  }
+
+  // Move from rhs. rhs.size() == 0 afterwards.
+  SizesAndStrides& operator=(SizesAndStrides&& rhs) {
+    if (this == &rhs) {
+      return *this;
+    }
+    if (C10_LIKELY(rhs.isInline())) {
+      if (C10_UNLIKELY(!isInline())) {
+        delete outOfLineStorage();
+      }
+      copyFromInline(rhs);
+    } else {
+      if (isInline()) {
+        // Steal their vector.
+        taggedStorageOrSize_ = rhs.taggedStorageOrSize_;
+      } else {
+        // Both out of line. Move their storage into ours.
+        *outOfLineStorage() = std::move(*rhs.outOfLineStorage());
+        delete rhs.outOfLineStorage();
```

Comment:
How come we can't just do `outOfLineStorage() = rhs.outOfLineStorage()`?

Reply:
OK, true, need to delete `outOfLineStorage()` first. But besides that?

Reply:
good point.